styx-common/src/main/java/com/spotify/styx/util/ShardedCounter.java

```diff
@@ -0,0 +1,241 @@
+          }
+          return null;
+        });
+      }
+    }
+
+    /**
+     * Gets the counter state. Creates it in Datastore if it doesn't exist yet.
+     */
+    private static CounterSnapshot fromDatastore(Datastore datastore, String counterId) {
+      CounterSnapshot snapshot = new CounterSnapshot();
+
+      final Key limitKey = datastore.newKeyFactory().setKind(KIND_COUNTER_LIMIT).newKey(counterId);
+      final Entity limit = datastore.get(limitKey);
+      if (limit == null) {
+        snapshot.limit = Long.MAX_VALUE;
+        // Or IllegalStateException("No limit found in Datastore for " + counterId);?
+      } else {
+        snapshot.limit = limit.getLong(PROPERTY_LIMIT);
+      }
+
+      final EntityQuery queryShards = EntityQuery.newEntityQueryBuilder()
+          .setKind(KIND_COUNTER_SHARD)
+          .setFilter(CompositeFilter.and(
+              PropertyFilter.ge("__key__", counterId + "-0"),
+              PropertyFilter.lt("__key__", counterId + "-" + NUM_SHARDS)))
+          .setOrderBy(OrderBy.asc("__key__"))
+          .setLimit(NUM_SHARDS)
+          .build();
+      final QueryResults<Entity> shards = datastore.run(queryShards);
+      snapshot.shards = new ArrayList<Long>(NUM_SHARDS);
+      int shardCount;
+      for (shardCount = 0; shards.hasNext(); shardCount++) {
+        Entity shard = shards.next();
+        String shardKey = shard.getKey().toString();
+        int shardIndex = Integer.valueOf(shardKey.substring(shardKey.indexOf('-') + 1));
+        snapshot.shards.set(shardIndex, shard.getLong(PROPERTY_SHARD));
+      }
+      if (shardCount < NUM_SHARDS) {
+        // The counter probably has not been initialized (so we have empty QueryResults). Also
+        // possible that a prior initialize() crashed halfway, or we got a partial list of shards in
+        // QueryResults due to eventual consistency. In any case, repeated initialization eventually
+        // creates all NUM_SHARDS shards.
+        initialize(datastore, counterId);
+      }
+
+      snapshot.updatedAt = Instant.now();
+      return snapshot;
+    }
+
+    private boolean isRecent() {
+      return updatedAt.plus(CACHE_EXPIRY_MILLIS, ChronoUnit.MILLIS).isAfter(Instant.now());
+    }
+
+    private long shardCapacity(int shardIndex) {
+      return limit / NUM_SHARDS + (shardIndex < limit % NUM_SHARDS ? 1 : 0);
+    }
+
+    /**
+     * Returns shard index which _likely_ could be successfully updated by delta, according to our
+     * cached view of the state in Datastore.
+     */
+    private int pickShardWithSpareCapacity(long delta) {
+      List<Integer> candidates = new ArrayList<>();
+
+      for (int i = 0; i < shards.size(); i++) {
+        if (shards.get(i) + delta >= 0
+            && shards.get(i) + delta <= shardCapacity(i)) {
+          candidates.add(i);
+        }
+      }
+
+      if (candidates.isEmpty()) {
+        return new Random().nextInt(NUM_SHARDS);
```

Comment:
At this point it looks like the counter does not have the capacity asked for and we might want to give that feedback to the user. If we just return random shard here we lose that information. Do something like throwing a `CounterFullException` to propagate this information to upper layers that can log it?

Reply:
We return a random choice here to later get the "Chosen shard has no spare capacity anymoar" exception (or succeed, with low likelihood), so upper layers do get that information.

(It looks like we'll just adopt the TransactionException from https://github.com/spotify/styx/pull/363 instead of ConcurrentModificationException.)

Reply:
Might throw that exception already here, but opted to do in just one place, to a slight perceived maintainability benefit.